### PR TITLE
Adding a continuous build and test using  a python selenium based runner.

### PR DESCRIPTION
**Pull request by mithro**
Features include;
- Auto-install of dependencies including needed python modules and needed selenium drivers.
- Ability to run against different browsers.
- Ability to use a Selenium grid, including support for Sauce Labs.
- Run in a headless mode by starting up VNC/xvfb - including uploading screenshots of failed tests to post-image.
- Ability to output subunit format, allowing integration with testrepository and similar tools.
- Runs on travis-ci

Things which need work
- Testing the code works on Windows and Mac OS X
- Use a better screenshot system.

### CODE REVIEW

File: .travis-setup.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	google-chrome --version
+	;;
+
+Firefox)
+	firefox --version
+	;;
+esac
+
+# FIXME: For speed we should install pip via apt-get install python-imaging
+#        then copy into the virtualenv.
```
Comment (shans) [inline .travis-setup.sh]: It looks like we do this?

File: run-tests.py [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+import argparse
+parser = argparse.ArgumentParser()
+parser.add_argument(
+    "-b", "--browser", type=str, required=True,
+    choices=['Firefox', 'Chrome', 'Ie', 'PhantomJS', 'Remote'],
+    help="Which WebDriver to use.")
+
+parser.add_argument(
+    "-x", "--virtual", action='store_true', default=False,
+    help="Use a virtual screen system such as Xvfb, Xephyr or Xvnc.")
```
Comment (shans) [inline run-tests.py:24]: Leave a blank line after this for consistency.

File: run-tests.py [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+    "-a", "--auto-install", action='store_true', default=True,
+    help="Auto install any dependencies.")
+
+parser.add_argument(
+    "-v", "--verbose", action='store_true', default=False,
+    help="Output more information.")
+
+parser.add_argument(
+    "-u", "--upload", action='store_true', default=False,
+    help="Upload images to picture sharing site, only really useful for testbots.")
```
Comment (shans) [inline run-tests.py]: Does this use a specific site? Maybe note which one here if so.

File: run-tests.py [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+
+# Only used by the Remote browser option.
+parser.add_argument(
+    "--remote-executor", type=str,
+    help="Location of the Remote executor.")
+parser.add_argument(
```
Comment (shans) [inline run-tests.py:48]: newline

File: run-tests.py [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+    "-s", "--sauce", action='store_true', default=False,
+    help="Use the SauceLab's Selenium farm rather then locally starting"
+         " selenium. SAUCE_USERNAME and SAUCE_ACCESS_KEY must be set in"
+         " environment.")
+
+# Subunit / testrepository support
+parser.add_argument(
+    "--subunit", action='store_true', default=False,
+    help="Output raw subunit binary data.")
+parser.add_argument(
```
Comment (shans) [inline run-tests.py:64]: newline

File: test/test-runner.html [hunk 6]
```diff
@@ -59,7 +62,13 @@
 }
 
 function postResultsToServer(resultObject) {
-  // TODO: post the results somewhere
+  var post_iframe = document.getElementById('result-post-frame');
+  post_iframe.onload = function(){
+      updateState(1);
+  };
+
+  post_iframe.contentDocument.getElementById('data').innerHTML = JSON.stringify(resultObject);
```
Comment (shans) [inline test/test-runner.html]: There is, alas, an 80 column limit in force for this project.

File: test/test-runner.html [hunk 7]
```diff
@@ -83,8 +103,7 @@
   document.querySelector('#test-link').href = src;
   document.querySelector('#test-frame').src = src;
   document.querySelector('#prev-button').disabled = currentTest <= 0;
-  document.querySelector('#next-button').disabled =
-      currentTest >= tests.length - 1;
+  document.querySelector('#next-button').disabled = currentTest >= tests.length - 1;
```
Comment (shans) [inline test/test-runner.html]: Keep under 80 cols.

**Comment by shans** [comment]
Looks good. You can merge this after responding to review comments.

File: .travis-setup.sh (hunk 1 above)
Comment (mithro) [inline .travis-setup.sh]: Done!

File: run-tests.py (hunk 2 above)
Comment (mithro) [inline run-tests.py:24]: Done.

File: run-tests.py (hunk 3 above)
Comment (mithro) [inline run-tests.py]: Done.

File: run-tests.py (hunk 4 above)
Comment (mithro) [inline run-tests.py:48]: Done.

File: run-tests.py (hunk 5 above)
Comment (mithro) [inline run-tests.py:64]: Done.

File: test/test-runner.html (hunk 6 above)
Comment (mithro) [inline test/test-runner.html]: Done.

File: test/test-runner.html (hunk 7 above)
Comment (mithro) [inline test/test-runner.html]: Done.

**Pull request merged**